Ajoute un bouton "signaler une erreur" sur les tutoriels

pierre-24 opened the pull request:
| Q | R |
| --- | --- |
| Correction de bugs ? | non |
| Nouvelle Fonctionnalité ? | oui |
| Tickets concernés | #1594, #1595 |

Soit un "objet", qui peut être un tuto (big ou mini) ou un chapitre. Cette PR vise à pouvoir signaler une erreur sur cet objet.
- Ajout d'une template `warn_typo.part.html`, qui prend pour argument `obj_type`, qui peut être soit `"chapter"`, soit `"tutorial"` et `obj_pk`, qui est le `pk` de l'objet, tel qu'il soit. Note: la variable `tutorial` doit également être définie, même si `obj_type` vaut `"tutorial"` (question de simplicité).
- Ajout d'une fonction `warn_typo()` dans `/zds/tutorial/view.html`, qui reprend en argument `obj_type` et `obj_pk` et envois un MP à tous les auteurs du tutoriel concernant l'objet en question (le tutoriel ou le chapitre du tutoriel). Le MP ne part pas si :
  - Un auteur s’envoie un MP à lui-même ;
  - La suggestion est vide ;
  - L'objet ne correspond à rien.

Le code a été écrit de telle manière à ce qu'on puisse éventuellement l'étendre: introduction, conclusion, parties, directement les extraits, ...
# Note de QA :
- Avec un user A, créer un big tutoriel. Lui rajouter une partie, puis un chapitre. Faire valider le tout.
- Avec un user B, signaler des erreurs sur le chapitre ET sur le tutoriel lui-même (actuellement, le bouton est en bas de la page). Vérifier que les MPs arrivent bien, vérifier qu'il n'est pas possible d'envoyer de messages vides. **Vérifier lque les liens sont justes.**
- Avec user A, vérifier qu'il n'est pas possible de s'envoyer des corrections: le bouton ne doit simplement pas apparaître;
- Faire de même avec un mini-tuto.
- Faire de même avec des tuto en bêta, en faisant bien attention que les liens dans le MP renvoient bien vers la version bêta (j'insiste)

pierre-24 committed:
Ajoute un bouton "signaler une erreur"

Soit un "objet", qui peut être un tuto (big ou mini) ou un chapitre.

- Ajout d'une template `warn_typo.part.html`, qui prend pour argument
  `obj_type`, qui peut etre soit `"chapter"`, soit `"tutorial"` et
  `obj_pk`, qui est le `pk` de l'objet, tel qu'il soit.
  Note: la variable `tutorial` doit egalement etre definie, même si
  `obj_type` vaut `"tutorial"` (question de simplicité).
- Ajout d'une fonction `warn_typo()` dans `/zds/tutorial/view.html`,
  qui reprend en argument `obj_type` et `obj_pk` et envois un MP
  à tous les auteurs du tutoriel concernant l'objet en question
  (le tutoriel ou le chapitre du tutoriel).
  Le MP ne part pas si :
  + Un auteur s'envoit un MP à lui-même ;
  + La suggestion est vide ;
  + L'objet ne correspond à rien.

gustavi commented on templates/tutorial/includes/warn_typo.part.html:
```diff
@@ -0,0 +1,38 @@
+<form class="modal modal-big" id="warn-typo-modal" action="{% url "zds.tutorial.views.warn_typo" obj_type obj_pk %}" method="post" >
+    <p>
+        {% if obj_type == "tutorial" %}
+                {% trans "J'ai trouvé une faute dans ce tutoriel" %}
```
L'indentation est étrange ici !

pierre-24 replied:
Un peu moins d'espace ? (genre 4 ?)

gustavi replied:
Oui, il me semble que c'est la règle !

Eskimon replied:
Yep, idem sur les quelques lignes au dessus... qqun c'est endormi sur sa barre espace ? :D

pierre-24 committed:
Merge branch 'dev' of https://github.com/zestedesavoir/zds-site into add_btn_error

pierre-24 commented:
J'ai tenu compte des remarques sur le code.

**GO QA** :)

pierre-24 committed:
Tien compte des remarques

- Indentation ;
- Renvoyer `Http404` is `obj_type` n'est pas bon ;
- Corriger la regle de redirection dans `url.py`

poulp commented:
J'ai pas QA, j'ai juste regarder le code mais j'ai quelques remarque : 
- ta vue permet d'envoyer une correction sur n'importe quel tuto sans distinction du fait qu'il soit en bêta ou pas, non publié, etc.
- Que dirais tu d'étendre ta pr pour toucher les articles ? Normalement il suffirait d'ajouter une possibilité de plus dans la vue et de mettre le bouton dans le template.

pierre-24 commented:
> ta vue permet d'envoyer une correction sur n'importe quel tuto sans distinction du fait qu'il soit en bêta ou pas, non publié, etc.

Pas faux, je vais limiter ça aux tutos publiés, comme c'est le cas pour l'apparition du bouton. 

> Que dirais tu d'étendre ta pr pour toucher les articles ? Normalement il suffirait d'ajouter une possibilité de plus dans la vue et de mettre le bouton dans le template.

En fait, c'est un peu plus "complexe" que ça, parce que tuto et articles sont actuellement pas géré dans la même vue.  En pratique, ça coûte pas grand chose de plus qu'un peu de copier-coller.

Ceci dit, j'ai peur d'une chose, qui est moins évidente que sur un article: quand il est publié, tout le monde ce "jette dessus" et l'auteur peut éventuellement être saturé de MP pour signaler la même erreur à chaque fois. Dans le cas d'un article, signaler l'erreur directement dans le flux de commentaire me parait plus intéréssant, parce que ça apporte de l'interaction (les gens réagissent, se corrigent, discutent) que dans un tuto, l'effet est plus limité. Mais je peux me tromper :)

poulp commented:
> Pas faux, je vais limiter ça aux tutos publiés, comme c'est le cas pour l'apparition du bouton

Prendre en compte les tutos en bêta aussi pourrait être sympa.

> En fait, c'est un peu plus "complexe" que ça, parce que tuto et articles sont actuellement pas géré dans la même vue. En pratique, ça coûte pas grand chose de plus qu'un peu de copier-coller.

oui vivement la zep07

pierre-24 committed:
Ajoute la possibilite d'avoir ce bouton en beta

pierre-24 commented:
Voilà, j'ai ajouté le bouton en bêta également et j'ai empêché l'user de donner son avis sur un draft.

Par contre, pour éviter de me disperser, je vais me limiter aux tutos dans cette PR, celle sur les articles devient alors _obvious_ :)

Eskimon commented:
Du coup on peut faire QA complete et definitive ici ?

pierre-24 commented:
Ah, juste, _my bad_, j'avais oublié cette PR là. Oui, on peut faire QA, manque que les TU et de la doc :)

Eskimon commented:
Je sais pas ou je me plante, mais je vois nulle part le bouton de signalement oO

pierre-24 commented:
En bas du tuto/chapitre, quand t'es pas auteur ;) (normalement :p )

Eskimon commented:
arf ma betise etant que je me baladais comme auteur :) j'y retourne !

Eskimon commented:
Pour moi tout semble ok...
- Le bouton est invisible pour un auteur du tuto
- On peut envoyer direct un message ou par MP pré-rempli
- Ok sur mini/big tuto
- Big tuto, ok sur "page général" ou chapitre

Bref, du beau boulot !

Pour l'allure du bouton, j'ai rien de mieux à proposer. J'ai envie de dire "mergeons et voyons les retours, il est pas moche et pas anti-ergonomique !"

Qu'en pensez-vous ?

pierre-24 commented:
Je suis d'accord, mais attend juste TU et doc pour merger (je vais essayer de voir si je peux pas faire ça ce soir).

pierre-24 committed:
Merge branch 'dev' of https://github.com/zestedesavoir/zds-site into add_btn_error

pierre-24 committed:
Ajout doc' et TU

pierre-24 commented:
Hop, dès que Travis est passé, pour moi c'est bon :)

Eskimon commented:
Beau boulot !
